test/dynamicppl/varinfo.jl

```diff
@@ -0,0 +1,369 @@
+
+using ..NumericalTests: check_numerical
+using AbstractPPL: VarName
+using BangBang: push!!, empty!!
+using LinearAlgebra: I
+using Test: @testset, @test
+using Turing
+
+@testset "varinfo.jl" begin
+    # Declare empty model to make the Sampler constructor work.
+    @model empty_model() = begin
+        x = 1
+    end
+
+    function randr(
+        vi::DynamicPPL.VarInfo, vn::VarName, dist::Distribution, spl::DynamicPPL.Sampler, count::Bool=false
+    )
+        if !haskey(vi, vn)
+            r = rand(dist)
+            push!!(vi, vn, r, dist, spl)
+            r
+        elseif DynamicPPL.is_flagged(vi, vn, "del")
+            DynamicPPL.unset_flag!(vi, vn, "del")
+            r = rand(dist)
+            vi[vn] = DynamicPPL.tovec(r)
+            DynamicPPL.setorder!(vi, vn, DynamicPPL.get_num_produce(vi))
+            r
+        else
+            count && checkindex(vn, vi, spl)
+            DynamicPPL.updategid!(vi, vn, spl)
+            vi[vn]
+        end
+    end
+
+    @testset "link!" begin
+        @model gdemo(x, y) = begin
+            s ~ InverseGamma(2, 3)
+            m ~ Uniform(0, 2)
+            x ~ Normal(m, sqrt(s))
+            y ~ Normal(m, sqrt(s))
+        end
+        model = gdemo(1.0, 2.0)
+
+        vi = DynamicPPL.VarInfo()
+        meta = vi.metadata
+        model(vi, DynamicPPL.SampleFromUniform())
+        @test all(x -> !DynamicPPL.istrans(vi, x), meta.vns)
+
+        alg = HMC(0.1, 5)
+        spl = DynamicPPL.Sampler(alg, model)
+        v = copy(meta.vals)
+        DynamicPPL.link!!(vi, spl, model)
+        @test all(x -> DynamicPPL.istrans(vi, x), meta.vns)
+        DynamicPPL.invlink!!(vi, spl, model)
+        @test all(x -> !DynamicPPL.istrans(vi, x), meta.vns)
+        @test meta.vals == v
+
+        vi = DynamicPPL.TypedVarInfo(vi)
+        meta = vi.metadata
+        alg = HMC(0.1, 5)
+        spl = DynamicPPL.Sampler(alg, model)
+        @test all(x -> !DynamicPPL.istrans(vi, x), meta.s.vns)
+        @test all(x -> !DynamicPPL.istrans(vi, x), meta.m.vns)
+        v_s = copy(meta.s.vals)
+        v_m = copy(meta.m.vals)
+        DynamicPPL.link!!(vi, spl, model)
+        @test all(x -> DynamicPPL.istrans(vi, x), meta.s.vns)
+        @test all(x -> DynamicPPL.istrans(vi, x), meta.m.vns)
+        DynamicPPL.invlink!!(vi, spl, model)
+        @test all(x -> !DynamicPPL.istrans(vi, x), meta.s.vns)
+        @test all(x -> !DynamicPPL.istrans(vi, x), meta.m.vns)
+        @test meta.s.vals == v_s
+        @test meta.m.vals == v_m
+
+        # Transforming only a subset of the variables
+        DynamicPPL.link!(vi, spl, Val((:m,)))
+        @test all(x -> !DynamicPPL.istrans(vi, x), meta.s.vns)
+        @test all(x -> DynamicPPL.istrans(vi, x), meta.m.vns)
+        DynamicPPL.invlink!(vi, spl, Val((:m,)))
+        @test all(x -> !DynamicPPL.istrans(vi, x), meta.s.vns)
+        @test all(x -> !DynamicPPL.istrans(vi, x), meta.m.vns)
+        @test meta.s.vals == v_s
+        @test meta.m.vals == v_m
```

Comment:
These methods are deprecated: 

https://github.com/TuringLang/DynamicPPL.jl/blob/f5890a11ef4002026e8edeb355b6e8a50c68d115/src/varinfo.jl#L1158-L1164

And AFAIK there is no replacement. The same is true of `invlink!`. Is it OK to drop these tests?

Reply:
@mhauru – maybe related to your work on Gibbs?

Reply:
Without testing out whether this works, maybe you could use `link!!(vi, sampler, model)` but with a sampler that has a space set, so e.g.
```julia
alg = HMC(0.1, 5, :m)
spl = DynamicPPL.Sampler(alg, model)
```
instead of `Val((:m,))` for `space`. I think it should have the same effect.

Eventually we would like to replace these with ones where you give the `VarName` to link as an argument, but I think we aren't there yet.

Reply:
This works!